app/services/dark-mode.ts

```diff
@@ -0,0 +1,135 @@
+import Service, { service } from '@ember/service';
+import { tracked } from '@glimmer/tracking';
+import { action } from '@ember/object';
+import { registerDestructor } from '@ember/destroyable';
+import type RouterService from '@ember/routing/router-service';
+import type LocalStorageService from 'codecrafters-frontend/services/local-storage';
+import RouteInfoMetadata from 'codecrafters-frontend/utils/route-info-metadata';
+
+export const LOCAL_STORAGE_KEY = 'dark-mode-preference';
+
+export type UpdateListener = (isEnabled: boolean) => void;
+
+export type LocalStoragePreference = 'system' | 'dark' | 'light';
+type SystemPreference = 'dark' | 'light';
+
+export default class DarkModeService extends Service {
+  @service declare router: RouterService;
+  @service declare localStorage: LocalStorageService;
+
+  /**
+   * Currently loaded localStorage preference value
+   */
+  @tracked localStoragePreference?: LocalStoragePreference;
+
+  /**
+   * Currently loaded system preference value
+   */
+  @tracked systemPreference?: SystemPreference;
+
+  /**
+   * Callback methods currently registered as listeners to Dark Mode environment changes
+   * @private
+   */
+  #registeredListeners: UpdateListener[] = [];
+
+  constructor(properties: object | undefined) {
+    super(properties);
+
+    // Load the localStorage preference from localStorage service
+    this.localStoragePreference = (this.localStorage.getItem(LOCAL_STORAGE_KEY) as LocalStoragePreference | null) || undefined;
+
+    // Create a media query to load current system Dark Mode preference
+    const mediaQuery = window.matchMedia('(prefers-color-scheme: dark)');
+
+    // Create a listener for media query result changes
+    const mqChangeListener = (e: MediaQueryListEvent) => {
+      this.systemPreference = e.matches ? 'dark' : 'light';
+      this.invokeUpdateListeners();
+    };
+
+    // Load the current system preference by executing the media query
+    this.systemPreference = mediaQuery.matches ? 'dark' : 'light';
+
+    // Subscribe to media query result changes
+    mediaQuery.addEventListener('change', mqChangeListener);
+
+    // Unsubscribe from media query changes when service is destroyed
+    registerDestructor(this, () => {
+      mediaQuery.removeEventListener('change', mqChangeListener);
+    });
+  }
+
+  /**
+   * Indicates whether Dark Mode is currently enabled: either because visiting a
+   * "dark route", or derived from localStorage & system preference
+   */
+  get isEnabled(): boolean {
+    return this.#isVisitingDarkRoute || this.#isLocalStorageEnabled;
+  }
+
+  /**
+   * Returns whether Dark Mode should be enabled based on localStorage & system preference
+   * @private
+   */
+  get #isLocalStorageEnabled(): boolean {
```

Comment:
Name is misleading - could be something like `isEnabledViaPreferences`

Reply:
Renamed to `#isEnabledViaPreferences`